pkg/diagnostics/networkpod/pod.go

```diff
@@ -0,0 +1,206 @@
+type CheckPodNetwork struct {
+	KubeClient *kclient.Client
+	OSClient   *osclient.Client
+}
+
+// Name is part of the Diagnostic interface and just returns name.
+func (d CheckPodNetwork) Name() string {
+	return CheckPodNetworkName
+}
+
+// Description is part of the Diagnostic interface and just returns the diagnostic description.
+func (d CheckPodNetwork) Description() string {
+	return "Check pod to pod communication in the cluster. In case of ovs-subnet network plugin, all pods should be able to communicate with each other and in case of multitenant network plugin, pods in non-global projects should be isolated and pods in global projects should be able to access any pod in the cluster and vice versa."
+}
+
+// CanRun is part of the Diagnostic interface; it determines if the conditions are right to run this diagnostic.
+func (d CheckPodNetwork) CanRun() (bool, error) {
+	if d.KubeClient == nil {
+		return false, errors.New("must have kube client")
+	} else if d.OSClient == nil {
+		return false, errors.New("must have openshift client")
+	}
+	return true, nil
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d CheckPodNetwork) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(CheckPodNetworkName)
+
+	pluginName, ok, err := util.GetOpenShiftNetworkPlugin(d.OSClient)
+	if err != nil {
+		r.Error("DPodNet1001", err, fmt.Sprintf("Checking network plugin failed. Error: %s", err))
+		return r
+	}
+	if !ok {
+		r.Warn("DPodNet1002", nil, fmt.Sprintf("Skipping pod connectivity test. Reason: Not using openshift network plugin."))
+		return r
+	}
+
+	localPods, nonlocalPods, err := util.GetLocalAndNonLocalPods(d.KubeClient)
+	if err != nil {
+		r.Error("DPodNet1003", err, fmt.Sprintf("Getting local and nonlocal pods failed. Error: %s", err))
+		return r
+	}
+
+	var vnidMap map[string]uint32
+	if sdnplugin.IsOpenShiftMultitenantNetworkPlugin(pluginName) {
+		netnsList, err := d.OSClient.NetNamespaces().List(kapi.ListOptions{})
+		if err != nil {
+			r.Error("DPodNet1004", err, fmt.Sprintf("Getting all network namespaces failed. Error: %s", err))
+			return r
+		}
+
+		vnidMap = map[string]uint32{}
+		for _, netns := range netnsList.Items {
+			vnidMap[netns.NetName] = netns.NetID
+		}
+	}
+
+	localGlobalPods, localNonGlobalPods := getGlobalAndNonGlobalPods(localPods, vnidMap)
+	nonlocalGlobalPods, nonlocalNonGlobalPods := getGlobalAndNonGlobalPods(nonlocalPods, vnidMap)
+
+	checkSameNodePodToPodConnection(localGlobalPods, localNonGlobalPods, vnidMap, r)
+	checkDifferentNodePodToPodConnection(localGlobalPods, localNonGlobalPods, nonlocalGlobalPods, nonlocalNonGlobalPods, vnidMap, r)
+	return r
+}
+
+func getGlobalAndNonGlobalPods(pods []kapi.Pod, vmap map[string]uint32) ([]kapi.Pod, []kapi.Pod) {
+	globalPods := []kapi.Pod{}
+	nonGlobalPods := []kapi.Pod{}
+
+	for _, pod := range pods {
+		if (vmap == nil) || (vmap != nil && vmap[pod.Namespace] == api.GlobalVNID) {
```

Comment:
`||` short-circuits, so you know `vmap != nil` in the second half.
also, if you assert that `api.GlobalVNID == 0` somewhere then you can simplify even further because `vmap[whatever]` will always return 0, so just

    if vmap[pod.Namespace] == api.GlobalVNID {

will do the right thing for both single- and multi-tenant